### PR TITLE
Feature/update marker style #162

### PR DESCRIPTION
**Pull request by Christopher-Sandbach**
Left colour mapping function without tests as I doubt it will be in the final system. Nonetheless, I can add them in if you think it worthwhile

### CODE REVIEW

**Comment by Christopher-Sandbach** [comment]
Basic colour mapping tests added as current build does not surpass test coverage threshold

**Comment by Christopher-Sandbach** [comment]
api.ts tests added to improve api test coverage

File: src/API/public/speciesList.json [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    {"species":"vaneedeni"},
+    {"species":"vanhoofi"},
+    {"species":"vernus"},
+    {"species":"vinckei"},
+    {"species":"walravensi"},
+    {"species":"wellcomei"},
+    {"species":"wilsoni"},
+    {"species":"ziemanni"}
+  ]
+}
```
Comment (andykitch) [inline src/API/public/speciesList.json:139]: Well done typing this all out, but... Can't we just get this list from the db? Then we won't have to maintain two sources of truth for the species list.

File: src/API/src/db/occurrence/occurrence.service.ts [hunk 2]
```diff
@@ -27,6 +29,7 @@ export class OccurrenceService {
       .orderBy('occurrence.id')
       .leftJoinAndSelect('occurrence.sample', 'sample')
       .leftJoinAndSelect('occurrence.site', 'site')
+      .leftJoinAndSelect('occurrence.recordedSpecies', 'recordedSpecies')
```
Comment (andykitch) [inline src/API/src/db/occurrence/occurrence.service.ts:32]: Does this actually get the species information out, or just the recordedSpecies?

File: src/API/src/db/shared/entities/recordedSpecies.resolver.test.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { buildTestingModule } from '../../../testHelpers';
```
Comment (andykitch) [inline src/API/src/db/shared/entities/recordedSpecies.resolver.test.ts:1]: Why is this test file here? It needs to be moved down a folder

File: src/API/src/db/shared/entities/species.entity.ts [hunk 4]
```diff
@@ -42,4 +43,15 @@ export class Species extends BaseEntity {
...
+  @OneToMany(
+    () => RecordedSpecies,
+    (recorded_species) => recorded_species.species,
+    {
+      eager: true,
+      cascade: true,
+      nullable: false,
+    },
+  )
+  recordedSpecies: RecordedSpecies[];
```
Comment (andykitch) [inline src/API/src/db/shared/entities/species.entity.ts:56]: I don't think we need this here - does it work without the recordedSpecies on the species entity?

File: src/API/src/db/shared/species.service.test.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      .mockResolvedValue(expectedSpecies);
+
+    const result = await service.findOneById('123');
+    expect(result).toEqual(expectedSpecies);
+    expect(speciesRepositoryMock.findOne).toHaveBeenCalledWith({
+      where: { id: '123' },
+    });
+  });
+
+  it('findAll returns all samples', async () => {
```
Comment (andykitch) [inline src/API/src/db/shared/species.service.test.ts:30]: samples again

File: src/API/src/db/occurrence/occurrence.service.ts (hunk 2 above)
Comment (Christopher-Sandbach) [reply src/API/src/db/occurrence/occurrence.service.ts:32]: I think it gets the species info out as it is available in the playground as a child of recordedSpecies, and then it's that species information which is used to colour the markers

File: src/API/src/ingest/occurrence.mapper.ts [hunk 6]
```diff
@@ -68,7 +68,9 @@ export const mapOccurrenceSite = (occurrence): Partial<Site> => {
   };
 };
 
-export const mapOccurrenceSpecies = (occurrence): Partial<RecordedSpecies> => {
+export const mapOccurrenceRecordedSpecies = (
```
Comment (andykitch) [inline src/API/src/ingest/occurrence.mapper.ts:71]: Can you fix the equivalent for the bionomics mapper please?

File: src/UI/components/map/map.tsx [hunk 7]
```diff
@@ -64,25 +66,27 @@ export const MapWrapper = () => {
...
-        text: 'Test text',
-        scale: 1.2,
-        fill: new Fill({
-          color: '#fff',
+    function markStyle(n_all: number, species: string) {
+      return new Style({
+        image: new Circle({
+          radius: 15,
+          fill: new Fill({
+            color: speciesColorMapRGB(speciesObject.data, species).color,
```
Comment (andykitch) [inline src/UI/components/map/map.tsx:74]: I assume this means the redux state has the `data` field? Can we deconstruct this before it's saved in the state, so we just save the species list?

File: src/UI/components/map/utils/speciesColorMapper.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  const speciesColorMapObject = arrayofSpeciesObjects.map(
+    (species: any, index: GLfloat) => ({
+      ...species,
+      color: hslToRgb(createHue(index, numSpecies), 1.0, 0.5),
+    })
+  );
+  return speciesColorMapObject.find((l: any) => l.species === speciesName);
+}
+
+export function hslToRgb(h: any, s: any, l: any) {
```
Comment (andykitch) [inline src/UI/components/map/utils/speciesColorMapper.ts:22]: Is there not a library that can do this for us? Looks like a lot of test cases will be needed to test this logic properly.

File: src/UI/components/map/utils/speciesColorMapper.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+export function speciesColorMapRGB(
+  arrayofSpeciesObjects: any,
+  speciesName: string
+) {
+  const numSpecies = arrayofSpeciesObjects.length;
+  const speciesColorMapObject = arrayofSpeciesObjects.map(
+    (species: any, index: GLfloat) => ({
+      ...species,
+      color: hslToRgb(createHue(index, numSpecies), 1.0, 0.5),
```
[the remark this answers is not in the document]
Comment (Christopher-Sandbach) [reply src/UI/components/map/utils/speciesColorMapper.ts:16]: Moved over to generating a hash based on species name. But as you mention distinguishing them could be an issue

File: src/UI/components/map/utils/speciesColorMapper.ts (hunk 8 above)
Comment (Christopher-Sandbach) [reply src/UI/components/map/utils/speciesColorMapper.ts:22]: Moved away from this method to using hash instead. If we return to this I will look around for an appropriate library

File: src/API/src/db/shared/entities/recordedSpecies.resolver.test.ts (hunk 3 above)
Comment (Christopher-Sandbach) [reply src/API/src/db/shared/entities/recordedSpecies.resolver.test.ts:1]: I just had a quick look in the docs as I was certain that that is where Ray had created it to begin with:

"We will create a central dependency injection module in a test helper, located in `src/API/src/testHelpers.ts` that provides a way of building mocked versions of the dependencies."

Is there a reason we would want it moved?

File: src/API/public/speciesList.json (hunk 1 above)
Comment (Christopher-Sandbach) [reply src/API/public/speciesList.json]: I'm unsure of the best way to approach this. When either yourself or Ray returns we can have a brief call to fill me in

File: src/UI/state/mapSlice.test.ts [hunk 10]
```diff
@@ -310,3 +311,15 @@ describe('getOccurrenceData', () => {
     );
   });
 });
+describe('drawerToggle', () => {
+  const { store } = mockStore({ map: initialState });
+  it('toggles the drawer from open to closed', () => {
+    const newState = reducer(initialState, drawerToggle());
+    expect(newState.map_drawer.open).toEqual(true);
```
Comment (andykitch) [inline src/UI/state/mapSlice.test.ts:318]: Can you add a test for the other way round? Closed to open?

File: src/API/src/db/shared/entities/recordedSpecies.resolver.test.ts (hunk 3 above)
Comment (andykitch) [reply src/API/src/db/shared/entities/recordedSpecies.resolver.test.ts:1]: Test files should always be at the same level as the file they're testing, unless there's a compelling reason not to do that.

File: src/UI/state/mapSlice.ts [hunk 11]
```diff
@@ -37,7 +37,7 @@ export interface MapState {
     baseMap: boolean;
   };
 
-  species_list: { species: string }[];
+  species_list: { series: string; color: number[] }[];
```
Comment (andykitch) [inline src/UI/state/mapSlice.ts:40]: I'm surprised (and slightly concerned) that this change hasn't made a test break in the slice tests... Can you change the test to reflect this change please?

**Review by andykitch — APPROVED** [review]

**Pull request merged**